Add initial example script for heateq with phase change

bgroenks96 opened the pull request:
This example demonstrates the solution of the 1D heat equation with mixed boundary conditions (Dirichlet upper, Neumann lower) and phase change effects included.

It uses the "free water" freezing characteristic which assumes that all phase change occurs at the the melting temperature (here 0 degrees C).

I think it's a nice test case for a simple yet also kind of nasty/stiff 1D PDE.

Currently, the gradient calculation seems to be roughly 1000x slower than the forward solve. Would be nice to improve this....? Maybe a better adjoint/VJP combination is possible?

facusapienza21 commented:
Hi @bgroenks96 !!! Thank you for contributing to the project! 

This is a very nice example that will be definitely useful for the review. Have you thought about ways in which we can use this example to expose differences between the different sensitivity methods? I think we can use this as one example for benchmarking different adjoints. What do you think? I think it will be interesting to see different ways of computing the gradient here, see how these compare in terms of accuracy (or at least check if these are the same), memory cost (total and maximum) and running time. 

Before merging your changes into `main`, I am not sure is a good idea to accept the `Manifest.toml` in the repository. I think the `Project.toml` should be enough and it is recommendable when the packages are going to be install in different machines and probably with other libraries too. Something that is on my list is to create a CI action for testing that all the code runs everytime a new change is added. Ideally, I would like to add this before accepting the PR, but it may be more efficient to move forward, integrate this and take care of this in the future. 

Great to have you onboard!!!

bgroenks96 commented:
>  I am not sure is a good idea to accept the Manifest.toml in the repository. 

You're right, my bad. I didn't realize that it got included. I removed it and added it to the `.gitignore`.

It would be great to do a comparison of different use VJPs/adjoint methods on a problem like this. I think the main issue is probably the number of time steps which blows up the memory allocation cost... but maybe there are also some optimizations that could be made.

bgroenks96 commented:
`InterpolatingAdjoint` with `EnzymeVJP` is 100x faster, so that's nice.
